File: app/src/main/java/h_mal/appttude/com/driver/Global/ImageSelectorResults.java
package h_mal.appttude.com.driver.Global;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.v4.app.ActivityCompat;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import static android.app.Activity.RESULT_OK;
import static android.os.Environment.DIRECTORY_PICTURES;
import static h_mal.appttude.com.driver.Global.ImageSelectorDialog.CAMERA_REQUEST;
import static h_mal.appttude.com.driver.Global.ImageSelectorDialog.PICK_IMAGE_REQUEST;
import static h_mal.appttude.com.driver.Global.ImageSelectorDialog.photoURI;
import static h_mal.appttude.com.driver.MainActivity.getDateStamp;

public class ImageSelectorResults {

    Activity activity;

    public interface FilepathResponse {
        void processFinish(Uri output);
    }

//    public FilepathResponse delegate;

    public ImageSelectorResults() {
    }

    public void Results(Activity activity, int requestCode, int resultCode, Intent data, Uri filePath,
                        ImageView imageView,FilepathResponse delegate){

        this.activity = activity;

        if(requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK
                && data != null && data.getData() != null )
        {
            filePath = data.getData();
            Bitmap bitmap = null;
            try {
                bitmap = MediaStore.Images.Media.getBitmap(activity.getContentResolver(), filePath);
                if (imageView.getVisibility() != View.VISIBLE) {
                    imageView.setVisibility(View.VISIBLE);
                }
                imageView.setImageBitmap(bitmap);
            }
            catch (IOException e)
            {
                e.printStackTrace();
            }
            finally {
                if (bitmap != null){
                    delegate.processFinish(filePath);
                    Log.i(getClass().getSimpleName(), "Results: " + filePath);
                }
            }


        }

        if (requestCode == CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            //check if we have we have storage rights
            int permission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE);

            if (permission != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(activity, "Storage permissions not granted", Toast.LENGTH_SHORT).show();
                return;
            }else {

                try {
                    File f = Environment.getExternalStoragePublicDirectory(DIRECTORY_PICTURES);
                    String fname = "driver"+ getDateStamp() + ".jpg";
                    File image = new File(f,fname);
                    FileOutputStream fileOutputStream = new FileOutputStream(image);

                    filePath = photoURI;

                    Bitmap bitmap = MediaStore.Images.Media
                            .getBitmap(activity.getContentResolver(), photoURI);

                    imageView.setImageBitmap(bitmap);

                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90,fileOutputStream);
                    fileOutputStream.flush();
                    fileOutputStream.close();
                    galleryAddPic();

                } catch (Exception e) {
                    e.printStackTrace();
                }

            }

            delegate.processFinish(filePath);
            Log.i(getClass().getSimpleName(), "Results: " + filePath);
        }
    }

    public void Results(Activity activity, int requestCode, int resultCode, Intent data, Uri filePath,FilepathResponse delegate){

        this.activity = activity;

        if(requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK
                && data != null && data.getData() != null )
        {
            filePath = data.getData();
            Bitmap bitmap = null;
            try {
                bitmap = MediaStore.Images.Media.getBitmap(activity.getContentResolver(), filePath);
            }
            catch (IOException e)
            {
                e.printStackTrace();
            }
            finally {
                if (bitmap != null){
                    delegate.processFinish(filePath);
                    Log.i(getClass().getSimpleName(), "Results: " + filePath);
                }
            }
        }

        if (requestCode == CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            //check if we have we have storage rights
            int permission = ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE);

            if (permission != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(activity, "Storage permissions not granted", Toast.LENGTH_SHORT).show();
                return;
            }else {

                try {
                    File f = Environment.getExternalStoragePublicDirectory(DIRECTORY_PICTURES);
                    String fname = "driver"+ getDateStamp() + ".jpg";
                    File image = new File(f,fname);
                    FileOutputStream fileOutputStream = new FileOutputStream(image);

                    filePath = photoURI;

                    Bitmap bitmap = MediaStore.Images.Media
                            .getBitmap(activity.getContentResolver(), photoURI);


                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90,fileOutputStream);
                    fileOutputStream.flush();
                    fileOutputStream.close();
                    galleryAddPic();

                } catch (Exception e) {
                    e.printStackTrace();
                }

            }

            delegate.processFinish(filePath);
            Log.i(getClass().getSimpleName(), "Results: " + filePath);

        }


    }

    private void galleryAddPic() {
        Intent mediaScanIntent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        File f = new File(photoURI.getPath());
        Uri contentUri = Uri.fromFile(f);
        mediaScanIntent.setData(contentUri);
        activity.sendBroadcast(mediaScanIntent);
    }

}

File: app/src/main/java/h_mal/appttude/com/driver/Global/ImageSelectorResults.kt
package h_mal.appttude.com.driver.Global

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Environment
import android.provider.MediaStore
import android.view.View
import android.widget.ImageView
import android.widget.Toast
import androidx.core.app.ActivityCompat
import h_mal.appttude.com.driver.utils.DateUtils.getDateStamp
import h_mal.appttude.com.driver.utils.DateUtils.getDateTimeStamp
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class ImageSelectorResults  //    public FilepathResponse delegate;
{
    lateinit var activity: Activity

    interface FilepathResponse {
        fun processFinish(output: Uri?)
    }

    fun Results(
        activity: Activity, requestCode: Int, resultCode: Int, data: Intent?, filePath: Uri?,
        imageView: ImageView?, delegate: FilepathResponse
    ) {
        var filePath: Uri? = filePath
        this.activity = activity
        if ((requestCode == ImageSelectorDialog.PICK_IMAGE_REQUEST) && (resultCode == Activity.RESULT_OK
                    ) && (data != null) && (data.data != null)
        ) {
            val uri = data.data
            var bitmap: Bitmap? = null
            try {
                bitmap =
                    MediaStore.Images.Media.getBitmap(activity.contentResolver, uri)
                if (imageView!!.visibility != View.VISIBLE) {
                    imageView.visibility = View.VISIBLE
                }
                imageView.setImageBitmap(bitmap)
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                if (bitmap != null) {
                    delegate.processFinish(uri)
                }
            }
        }
        if (requestCode == ImageSelectorDialog.CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            //check if we have we have storage rights
            val permission: Int = ActivityCompat.checkSelfPermission(
                (activity)!!,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            )
            if (permission != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(activity, "Storage permissions not granted", Toast.LENGTH_SHORT)
                    .show()
                return
            } else {
                try {
                    val f: File =
                        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
                    val fname: String = "driver" + getDateStamp() + ".jpg"
                    val image = File(f, fname)
                    val fileOutputStream = FileOutputStream(image)
                    filePath = ImageSelectorDialog.photoURI
                    val bitmap: Bitmap = MediaStore.Images.Media
                        .getBitmap(
                            activity.contentResolver,
                            ImageSelectorDialog.photoURI
                        )
                    imageView!!.setImageBitmap(bitmap)
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90, fileOutputStream)
                    fileOutputStream.flush()
                    fileOutputStream.close()
                    galleryAddPic()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            delegate.processFinish(filePath)
        }
    }

    fun Results(
        activity: Activity,
        requestCode: Int,
        resultCode: Int,
        data: Intent?,
        filePath: Uri?,
        delegate: FilepathResponse
    ) {
        var filePath: Uri? = filePath
        this.activity = activity
        if ((requestCode == ImageSelectorDialog.PICK_IMAGE_REQUEST) && (resultCode == Activity.RESULT_OK
                    ) && (data != null) && (data.data != null)
        ) {
            filePath = data.data
            var bitmap: Bitmap? = null
            try {
                bitmap =
                    MediaStore.Images.Media.getBitmap(activity.contentResolver, filePath)
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                if (bitmap != null) {
                    delegate.processFinish(filePath)
                }
            }
        }
        if (requestCode == ImageSelectorDialog.CAMERA_REQUEST && resultCode == Activity.RESULT_OK) {
            //check if we have we have storage rights
            val permission: Int = ActivityCompat.checkSelfPermission(
                (activity),
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            )
            if (permission != PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(activity, "Storage permissions not granted", Toast.LENGTH_SHORT)
                    .show()
                return
            } else {
                try {
                    val f: File =
                        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
                    val fname: String = "driver" + getDateTimeStamp() + ".jpg"
                    val image: File = File(f, fname)
                    val fileOutputStream: FileOutputStream = FileOutputStream(image)
                    filePath = ImageSelectorDialog.photoURI
                    val bitmap: Bitmap = MediaStore.Images.Media
                        .getBitmap(
                            activity.contentResolver,
                            ImageSelectorDialog.photoURI
                        )
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90, fileOutputStream)
                    fileOutputStream.flush()
                    fileOutputStream.close()
                    galleryAddPic()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            delegate.processFinish(filePath)
        }
    }

    private fun galleryAddPic() {
        val mediaScanIntent: Intent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
        val f: File = File(ImageSelectorDialog.photoURI!!.path)
        val contentUri: Uri = Uri.fromFile(f)
        mediaScanIntent.data = contentUri
        activity.sendBroadcast(mediaScanIntent)
    }
}